### ISSUE TITLE
Built-in Column Encryption Capability

### ISSUE THREAD

**Issue by pdevito3** [dotnet/efcore#31832]
## Problem
As a developer, if i want to encrypt a column in my database, I have to handroll the db integration.

## Possible Solution/Api
I'm no expert in this area and I know the dotnet team has done something similar to this under the hood for something like Identity, so maybe y'all have some thoughts here, but as a consumer, something like this might work:

```csharp
public sealed class PatientConfiguration : IEntityTypeConfiguration<Patient>
{
    private readonly IDataProtectionProvider _dataProtectionProvider;

    public PatientConfiguration(IDataProtectionProvider dataProtectionProvider)
    {
        _dataProtectionProvider = dataProtectionProvider;
    }

    public void Configure(EntityTypeBuilder<Patient> builder)
    {
        builder.Property(x => x.Name)
            .HasEncryption(_dataProtectionProvider)
    }
}
```

A simple example of this under the hood might look like this if working with postgres, butthis would need something more flexible/robust for the lib option

```csharp
public static class EncryptionExtensions
{
    public static PropertyBuilder<TProperty> HasEncryption<TProperty>(this PropertyBuilder<TProperty> propertyBuilder, IDataProtectionProvider dataProtectionProvider)
        => propertyBuilder.HasColumnType("bytea")
            .HasConversion(new ColumnEncryptor<TProperty>(dataProtectionProvider));
}

public class ColumnEncryptor : ColumnEncryptor<string>
{
    public ColumnEncryptor(IDataProtectionProvider dataProtectionProvider) : base(dataProtectionProvider) { }
}

public class ColumnEncryptor<TProperty> : ValueConverter<TProperty, byte[]>
{
    public ColumnEncryptor(IDataProtectionProvider dataProtectionProvider) :
        base(
            x => ProtectAndSerialize(x, dataProtectionProvider),
            x => DeserializeAndUnprotect(x, dataProtectionProvider),
            default
        )
    {
    }

    private static byte[] ProtectAndSerialize(TProperty x, IDataProtectionProvider dataProtectionProvider)
    {
        return dataProtectionProvider.CreateProtector("EncryptedColumn")
            .Protect(JsonSerializer.SerializeToUtf8Bytes(x));
    }

    private static TProperty DeserializeAndUnprotect(byte[] x, IDataProtectionProvider dataProtectionProvider)
    {
        var unprotectedData = dataProtectionProvider.CreateProtector("EncryptedColumn")
            .Unprotect(x);

        var deserializedData = JsonSerializer.Deserialize<TProperty>(unprotectedData);

        if (deserializedData == null)
        {
            throw new InvalidOperationException("Failed to deserialize the data for encryption.");
        }

        return deserializedData;
    }
}
```

**Comment by roji** [comment]
@pdevito3 I don't think EF should go into the business of defining something like an IDataProtectionProvider. Encryption is a vast domain with many possibilities, and an abstraction like this definitely doesn't belong in a data access layer/ORM. EF does provider the value converter infrastructure which you can use to encrypt/decrypt column data, but how exactly that's done (which algorithm, how exactly the key(s) are managed) is something that should be left outside. Of course, people are free to build an extension on top of EF that make column encryption easier (making use of value converters).

Note that this question also has some important database-specific aspects. For example, SQL Server has some built-in features around column encryption (e.g. [docs](https://learn.microsoft.com/en-us/sql/relational-databases/security/encryption/encrypt-a-column-of-data?view=sql-server-ver16)), so a general EF solution wouldn't make sense.

**Comment by pdevito3** [comment]
Well, I don't think it would have to prescribe how you encrypt or even that you use a data protection instance at all really. 

Even the simple example I showed doesn't prescribe any of what your mentioned. It just uses my own data protection provider and could easily be replaced by a bespoke function too.

Regardless, I hear you on a larger scale. I didn't see any discussions on this in the history so figured I'd raise it and see thoughts from the team and at least have a basic reference if other came searching.

**Comment by roji** [comment]
@pdevito3 if the actual data protection provider (i.e. IDataProtectionProvider) must be provided by the user, then there's very little value being added here no? All the user has to do is wire up that provider via a value converter?

**Comment by pdevito3** [comment]
> if the actual data protection provider (i.e. IDataProtectionProvider) must be provided by the user, then there's very little value being added here no? All the user has to do is wire up that provider via a value converter?

Maybe, but at least it would provide a standard abstraction and guidance for how it works.

In a perfect world, if it could be optionally taken a step further and have EF have a default implementation that would do encryption for you I'm sure that would help a lot but that's probably much less practical as already discussed.

**Comment by roji** [comment]
> Maybe, but at least it would provide a standard abstraction and guidance for how it works.

I think the only meaningful abstraction here is the value converter itself. If you introduce an extra abstraction whose only value is that users must provide an IDataProtectionProvider implementation rather than wiring up logic into a value converter, you've not really improved the developer experience in any way - the developer basically has to do the same thing in both cases.

Designing an IDataProtectionProvider abstraction would also be far from trivial - presumably it requires a key, but what format key should that be? Is something else maybe required? Encryption systems and algorithms vary considerably, and an abstraction over that would need to provide substantial value, which I really am not seeing here.

> In a perfect world, if it could be optionally taken a step further and have EF have a default implementation that would do encryption for you I'm sure that would help a lot but that's probably much less practical as already discussed.

I really wouldn't feel comfortable with EF making a decision on what a reasonable encryption default would be (at which point it's also EF's responsibility to update that choice if the default changes for some reason, deal with the breaking changes resulting from such a change...).

Bottom line, if you're interested in column encryption as a user, you need to understand your specific database's capabilities in this area, and be involved in the actual security-related choices related to that.

**Comment by pdevito3** [comment]
Yup makes sense. Appreciate the discussion. Hopefully this helps some others coming across this in the future 🍻

**Comment by roji** [comment]
Sure thing, and thanks for engaging! We also learn tons from these conversations, regardless what the outcome is.
